File: itanna/emacs/venv.py
"""
Virtual Environment Detection for Emacs
=========================================

Provides the canonical path to the Itanna Poetry venv so Emacs
can activate it automatically when opening the distribution.

Emacs will call:
    python3 -m itanna.emacs.venv path
    python3 -m itanna.emacs.venv python  (returns python binary path)
    python3 -m itanna.emacs.venv info    (returns JSON blob)
"""

import json
import os
import subprocess
import sys
from pathlib import Path


def _find_poetry_venv() -> Path | None:
    """Locate the Poetry-managed virtual environment for this project.

    Checks:
      1. VIRTUAL_ENV environment variable
      2. .venv/ inside the project root
      3. Poetry's cache directory ~/Library/Caches/pypoetry/virtualenvs/
    """
    # 1. Already activated
    venv = os.environ.get("VIRTUAL_ENV")
    if venv:
        return Path(venv)

    # 2. In-project venv
    project_root = Path(__file__).resolve().parent.parent.parent
    in_project = project_root / ".venv"
    if in_project.exists() and (in_project / "bin" / "python").exists():
        return in_project

    # 3. Poetry cache
    try:
        result = subprocess.run(
            ["poetry", "env", "info", "--path"],
            capture_output=True,
            text=True,
            cwd=project_root,
            timeout=10,
        )
        if result.returncode == 0:
            path = result.stdout.strip()
            if path:
                return Path(path)
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass

    return None


def _get_project_root() -> Path:
    """Return the itanna project root directory."""
    return Path(__file__).resolve().parent.parent.parent


def cmd_path() -> None:
    """Print the venv path, or nothing if not found."""
    venv = _find_poetry_venv()
    if venv:
        print(venv)
    else:
        print("", end="")


def cmd_python() -> None:
    """Print the path to the venv's python binary."""
    venv = _find_poetry_venv()
    if venv:
        py = venv / "bin" / "python"
        if py.exists():
            print(py)
            return
    print(sys.executable)


def cmd_info() -> None:
    """Print a JSON blob with environment info."""
    venv = _find_poetry_venv()
    info = {
        "project_root": str(_get_project_root()),
        "venv_path": str(venv) if venv else None,
        "python": str(venv / "bin" / "python") if venv and (venv / "bin" / "python").exists() else sys.executable,
        "active": os.environ.get("VIRTUAL_ENV") is not None,
        "in_project_venv": (Path(_get_project_root()) / ".venv").exists(),
    }
    print(json.dumps(info, indent=2))


def cmd_activate_script() -> None:
    """Print a shell script that activates the venv.

    Emacs can source this or use it to set environment variables.
    """
    venv = _find_poetry_venv()
    if not venv:
        print("echo 'No Itanna venv found. Run: poetry install'")
        return

    bin_dir = venv / "bin"
    print(f"""# Itanna venv activation snippet (sourced by Emacs)
export VIRTUAL_ENV="{venv}"
export PATH="{bin_dir}:$PATH"
export PYTHONPATH="{_get_project_root()}:${{PYTHONPATH:-}}"
""")

    # Also print individual exports for Emacs's process-environment
    print(f"__ITANNA_VENV={venv}")
    print(f"__ITANNA_PYTHON={bin_dir / 'python'}")


if __name__ == "__main__":
    commands = {
        "path": cmd_path,
        "python": cmd_python,
        "info": cmd_info,
        "activate": cmd_activate_script,
    }
    cmd = sys.argv[1] if len(sys.argv) > 1 else "info"
    handler = commands.get(cmd, cmd_info)
    handler()

File: vscode-itanna/src/venv.ts
// ⚡ Itanna Virtual Environment Activation
// Detects and activates the Poetry-managed Python virtual environment
// for the Itanna workspace, updating PATH and environment variables.

import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import { execSync, exec } from 'child_process';

let venvActivated = false;

/**
 * Find the Itanna VSCode virtual environment path.
 *
 * Priority:
 * 1. VIRTUAL_ENV environment variable
 * 2. `~/.itanna/.venv/` (dedicated VSCode environment)
 * 3. `.venv/` inside workspace root
 * 4. Poetry's cache directory
 *
 * The dedicated VSCode venv at `~/.itanna/.venv` is preferred because:
 * - It's independent of the Emacs/org-babel venv
 * - It persists regardless of which directory VSCode is opened in
 * - It's managed by the VSCode extension install script
 */
export async function findVenvPath(workspaceRoot: string): Promise<string | null> {
  // 1. Already activated in shell
  const envVar = process.env.VIRTUAL_ENV;
  if (envVar && fs.existsSync(envVar)) {
    return envVar;
  }

  // 2. Dedicated VSCode venv at ~/.itanna/.venv (highest priority for VSCode)
  const home = process.env.HOME || process.env.USERPROFILE || '';
  const itannaDir = path.join(home, '.itanna');
  const itannaVenv = path.join(itannaDir, '.venv');
  const itannaPython = path.join(itannaVenv, 'bin', 'python');
  if (fs.existsSync(itannaPython)) {
    return itannaVenv;
  }

  // 3. In-project venv (shared with Emacs)
  const inProjectVenv = path.join(workspaceRoot, '.venv');
  const inProjectPython = path.join(inProjectVenv, 'bin', 'python');
  if (fs.existsSync(inProjectPython)) {
    return inProjectVenv;
  }

  // 4. Poetry cache (async)
  try {
    const poetryPath = await execAsync('poetry', ['env', 'info', '--path'], { cwd: workspaceRoot });
    const cachePath = poetryPath.trim();
    if (cachePath && fs.existsSync(cachePath)) {
      return cachePath;
    }
  } catch {
    // poetry not available or not configured
  }

  return null;
}

/**
 * Ensure the Itanna Jupyter kernel is registered.
 * This creates a kernel spec so the Itanna venv appears as
 * "Python 3 (Itanna)" in VSCode's Jupyter kernel picker.
 */
async function ensureItannaKernel(pythonBin: string, verbose: boolean): Promise<void> {
  try {
    // Check if ipykernel is available
    const checkResult = execSync(
      `"${pythonBin}" -c "import ipykernel; print('ok')" 2>/dev/null`,
      { encoding: 'utf-8', timeout: 5000 }
    ).trim();

    if (checkResult !== 'ok') {
      if (verbose) {
        console.warn('ipykernel not available in venv. Install it: pip install ipykernel');
      }
      return;
    }

    // Check if the itanna kernel is already registered
    const existingKernel = execSync(
      `"${pythonBin}" -m jupyter kernelspec list 2>/dev/null`,
      { encoding: 'utf-8', timeout: 5000 }
    );

    if (existingKernel.includes('itanna ')) {
      if (verbose) console.log('✓ Itanna Jupyter kernel already registered');
      return;
    }

    // Register the kernel
    const result = execSync(
      `"${pythonBin}" -m ipykernel install --user --name itanna --display-name "Python 3 (Itanna)" 2>&1`,
      { encoding: 'utf-8', timeout: 10000 }
    ).trim();

    console.log(`✓ Itanna Jupyter kernel registered: ${result}`);

    // Notify the user
    if (vscode.window) {
      vscode.window.showInformationMessage(
        '⚡ Itanna Jupyter kernel registered. You can now select "Python 3 (Itanna)" in your notebooks.'
      );
    }
  } catch (err) {
    if (verbose) console.warn('Failed to register Itanna Jupyter kernel:', err);
  }
}

/**
 * Derive the Itanna project root from a venv path.
 *
 * The venv can be at:
 *   - <itanna-project>/.venv/          (project-local)
 *   - ~/.itanna/.venv/                 (dedicated VSCode venv)
 *
 * In the second case, the itanna project root is still the original
 * project directory (e.g., ~/projects/itanna/).
 */
function deriveItannaRoot(venvPath: string): string | undefined {
  const parent = path.resolve(venvPath, '..');

  // Case 1: parent is the project root (project-local venv)
  if (
    fs.existsSync(path.join(parent, 'pyproject.toml')) &&
    fs.existsSync(path.join(parent, 'templates-jupyter'))
  ) {
    return parent;
  }

  // Case 2: parent is ~/.itanna (dedicated VSCode venv)
  // Look for the actual project root in common locations
  const home = process.env.HOME || '';
  const commonRoots = [
    path.join(home, 'projects', 'itanna'),
    path.join(home, 'itanna'),
    path.join(home, 'src', 'itanna'),
  ];
  for (const root of commonRoots) {
    if (
      fs.existsSync(path.join(root, 'pyproject.toml')) &&
      fs.existsSync(path.join(root, 'templates-jupyter'))
    ) {
      return root;
    }
  }

  return undefined;
}

/**
 * Activate the Itanna virtual environment.
 * Updates relevant environment variables for the extension host.
 * Returns true if successful.
 */
export async function activateVenv(verbose: boolean = false): Promise<boolean> {
  if (venvActivated) {
    return true;
  }

  const workspaceRoot = vscode.workspace.workspaceFolders?.[0]?.uri.fsPath;

  try {
    let venvPath: string | null = null;
    let itannaRoot: string | undefined;

    // 1. Try to find venv relative to workspace root
    if (workspaceRoot) {
      venvPath = await findVenvPath(workspaceRoot);
    }

    // 2. If not found, try known itanna directories
    if (!venvPath) {
      const home = process.env.HOME || '';
      const knownRoots = [
        path.join(home, 'projects', 'itanna'),
        path.join(home, 'itanna'),
        process.env.ITANNA_ROOT || '',
      ];
      for (const root of knownRoots) {
        if (root) {
          venvPath = await findVenvPath(root);
          if (venvPath) {
            itannaRoot = root;
            break;
          }
        }
      }
    } else {
      // Derive itanna root from venv path
      itannaRoot = deriveItannaRoot(venvPath);
    }

    if (!venvPath) {
      if (verbose) {
        vscode.window.showWarningMessage(
          'Itanna virtual environment not found. Run `poetry install` in the terminal.'
        );
      }
      return false;
    }

    // If we couldn't derive itanna root from venv, use workspace root
    if (!itannaRoot) {
      itannaRoot = workspaceRoot || undefined;
    }

    const pythonBin = path.join(venvPath, 'bin', 'python');
    if (!fs.existsSync(pythonBin)) {
      return false;
    }

    // Set environment variables for the extension host
    process.env.VIRTUAL_ENV = venvPath;
    const binDir = path.join(venvPath, 'bin');
    const currentPath = process.env.PATH || '';
    if (!currentPath.includes(binDir)) {
      process.env.PATH = `${binDir}:${currentPath}`;
    }
    const safeRoot = itannaRoot || workspaceRoot || '';
    process.env.PYTHONPATH = [
      safeRoot,
      path.join(safeRoot, 'electrical'),
      path.join(safeRoot, 'itanna'),
      process.env.PYTHONPATH || ''
    ].filter(Boolean).join(':');
    process.env.ITANNA_ROOT = safeRoot;

    // Also set for the VS Code Python extension
    const pythonConfig = vscode.workspace.getConfiguration('python');
    await pythonConfig.update(
      'defaultInterpreterPath',
      pythonBin,
      vscode.ConfigurationTarget.Workspace
    );

    // Verify activation
    const version = execSync(`"${pythonBin}" --version`, { encoding: 'utf-8' }).trim();
    console.log(`⚡ Itanna venv activated: ${venvPath} (${version})`);

    // Ensure Jupyter kernel is registered
    await ensureItannaKernel(pythonBin, verbose);

    venvActivated = true;
    return true;
  } catch (err) {
    console.error('Failed to activate Itanna venv:', err);
    return false;
  }
}

/**
 * Deactivate the virtual environment (reset env vars).
 */
export function deactivateVenv(): void {
  delete process.env.VIRTUAL_ENV;
  // Note: we can't easily undo PATH modifications, but the extension host
  // lifetime is tied to the VSCode window, so this is fine.
  venvActivated = false;
}

/**
 * Check if the Itanna venv is currently activated.
 */
export function isVenvActive(): boolean {
  return venvActivated || !!process.env.VIRTUAL_ENV;
}

/**
 * Get the Python binary path (venv or system).
 */
export function getPythonPath(): string {
  if (process.env.VIRTUAL_ENV) {
    const venvPython = path.join(process.env.VIRTUAL_ENV, 'bin', 'python');
    if (fs.existsSync(venvPython)) {
      return venvPython;
    }
  }
  return 'python3';
}

/**
 * Helper: run a command and return stdout.
 */
function execAsync(cmd: string, args: string[], options: { cwd?: string }): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = exec(
      `${cmd} ${args.join(' ')}`,
      { cwd: options.cwd, encoding: 'utf-8', timeout: 10000 },
      (err, stdout, stderr) => {
        if (err) reject(err);
        else resolve(stdout);
      }
    );
  });
}
